File: DataStructures/python/linkedList/doublyLinked.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class List:
    def __init__(self):
        self.size = 0
        self.head = None
        self.tail = None

    def add(self, data):
        element = Node(data)

        if self.head is None:
            self.head = self.tail = element
        else:
            element.prev = self.tail
            self.tail.next = element
            self.tail = element

        self.size += 1

    def remove(self, data):
        if self.size == 0:
            return

        element = self.search(data)

        if element is None:
            return

        if element == self.head:
            if element == self.tail:
                self.tail = self.head = None
            else:
                self.head = element.next

            if self.head != None:
                self.head.prev = None

        else:
            if element == self.tail:
                self.tail = element.prev

            element.prev.next = element.next

        self.size -= 1

    def search(self, data):
        if self.size == 0:
            return None

        current = self.head
        while current != None:
            if current.data == data:
                break

            current = current.next
        
        return current

    def __str__(self):
        elements = self.getElementsData()

        return ', '.join(elements)

    def getElementsData(self):
        nodes = []

        cur = self.head
        while cur != None:
            nodes.append(cur.data)
            cur = cur.next

        return nodes

File: golang/linkedList/doublyLinked.go
package main

import (
	"fmt"
	"log"
)

type Node struct {
	Value int
	Prev  *Node
	Next  *Node
}

type List struct {
	size int
	head *Node
	tail *Node
}

func (l *List) Size() int {
	return l.size
}

func (l *List) Add(val int) error {
	if l.Find(val) != nil {
		return fmt.Errorf("Element already exists.")
	}

	node := Node{Value: val}

	if l.head == nil {
		l.head = &node
	} else {
		l.tail.Next = &node
		node.Prev = l.tail
	}

	l.tail = &node
	l.size++

	return nil
}

func (l *List) Remove(val int) error {
	node := l.Find(val)

	if node == nil {
		return fmt.Errorf("Element not found.")
	}

	if node.Prev == nil {
		l.head = node.Next
	} else {
		node.Prev.Next = node.Next
		node.Prev = nil
	}

	if node.Next == nil {
		l.tail = node.Prev
	} else {
		node.Next.Prev = node.Prev
		node.Next = nil
	}

	l.size--

	return nil
}

func (l *List) Find(val int) *Node {
	cur := l.head
	for cur != nil {
		if cur.Value == val {
			return cur
		}

		tmp := cur
		cur = tmp.Next
	}

	return nil
}

func (l *List) PrintAll() {
	log.Println("Size: ", l.size)

	cur := l.head
	for cur != nil {
		log.Println("Value:", cur.Value, "Prev:", cur.Prev, "Next:", cur.Next)

		tmp := cur
		cur = tmp.Next
	}
}
